xds/internal/xdsclient/e2e_test/federation_watchers_test.go

```diff
@@ -0,0 +1,403 @@
+)
+
+const testNonDefaultAuthority = "non-default-authority"
+
+// setupForFederationWatchersTest spins up two management servers, one for the
+// default (empty) authority and another for a non-default authority.
+//
+// Returns the management server, the nodeID to use, and the xDS client.
+func setupForFederationWatchersTest(t *testing.T) (*e2e.ManagementServer, string, xdsclient.XDSClient) {
+	overrideFedEnvVar(t)
+
+	// Start a management server as the default authority.
+	serverDefaultAuthority, err := e2e.StartManagementServer(nil)
+	if err != nil {
+		t.Fatalf("Failed to spin up the xDS management server: %v", err)
+	}
+	t.Cleanup(serverDefaultAuthority.Stop)
+
+	// Start another management server as the other authority.
+	serverTestAuthority, err := e2e.StartManagementServer(nil)
+	if err != nil {
+		t.Fatalf("Failed to spin up the xDS management server: %v", err)
+	}
+	t.Cleanup(serverTestAuthority.Stop)
+
+	// Create a bootstrap file in a temporary directory.
+	nodeID := uuid.New().String()
+	bootstrapContents, err := xdsinternal.BootstrapContents(xdsinternal.BootstrapOptions{
+		Version:                            xdsinternal.TransportV3,
+		NodeID:                             nodeID,
+		ServerURI:                          serverDefaultAuthority.Address,
+		ServerListenerResourceNameTemplate: e2e.ServerListenerResourceNameTemplate,
+		// Specify the address of the non-default authority.
+		Authorities: map[string]string{testNonDefaultAuthority: serverTestAuthority.Address},
+	})
+	if err != nil {
+		t.Fatalf("Failed to create bootstrap file: %v", err)
+	}
+	// Create an xDS client with the above bootstrap contents.
+	client, err := xdsclient.NewWithBootstrapContentsForTesting(bootstrapContents)
+	if err != nil {
+		t.Fatalf("Failed to create xDS client: %v", err)
+	}
+	return serverTestAuthority, nodeID, client
+}
+
+// TestFederation_ListenerResourceContextParamOrder covers the case of watching
+// a Listener resource with the new style resource name and context parameters.
+// The test verifies the following:
+// - Two watches with the same query string, but in different order results in
+//   the two watches watching the same resource.
+// - Response with the same query string, but in different order results in both
+//   watches being notified.
+func (s) TestFederation_ListenerResourceContextParamOrder(t *testing.T) {
+	serverTestAuthority, nodeID, client := setupForFederationWatchersTest(t)
+	defer client.Close()
+
+	var (
+		// Two resource names only differ in context parameter __order__.
+		resourceName1 = fmt.Sprintf("xdstp://%s/envoy.config.listener.v3.Listener/xdsclient-test-lds-resource?a=1&b=2", testNonDefaultAuthority)
+		resourceName2 = fmt.Sprintf("xdstp://%s/envoy.config.listener.v3.Listener/xdsclient-test-lds-resource?b=2&a=1", testNonDefaultAuthority)
+	)
+
+	// Register two watches for the listener resource names with the same query
+	// string, but in different order.
```

Comment:
Just a reminder "in" here and elsewhere please switch.

Reply:
Also, these comments map to all the xDS ordering tests, Listener, Route, Cluster, and Endpoint.

Reply:
Done.